### PR TITLE
[AIRFLOW-2744] Allow RBAC to accept plugins for views and links.

### PR DESCRIPTION
**Pull request by oliviersm199**
### Description

Airflow Users that wish to create plugins for the new www_rbac UI can not add plugin views or links. This PR fixes that by letting a user specify their plugins for www_rbac and maintains backwards compatibility with the existing plugins system.

Adds two additional members of an AirflowPlugin class to cover role based views and role based menu items. Specifies that the user can create a dictionary which has the metadata necessary to create a FlaskAppBuilder BaseView or MenuItem. This lets the plugins manager extend to the new www_rbac GUI.

### Tests

Adds unit tests under the plugins manager to test that the www_rbac views and menu links get added successfully.

### Documentation

Demonstrates in the plugins page how a user would create plugins for the new UI with an example.

### CODE REVIEW

File: airflow/plugins_manager.py [hunk 1]
```diff
@@ -49,6 +49,8 @@ class AirflowPlugin(object):
     admin_views = []
     flask_blueprints = []
     menu_links = []
+    role_based_views = []
+    role_based_menu_items = []
```
Comment (ashb) [inline airflow/plugins_manager.py]: Lets call these `appbuilder_views` etc. The "role" is just the reason we switched to FAB.

**Review by ashb — COMMENTED** [review]
Nice!

File: airflow/plugins_manager.py (hunk 1 above)
Comment (oliviersm199) [reply airflow/plugins_manager.py]: Changed it to use appbuilder_views etc.

**Comment by oliviersm199** [comment]
not sure why the code coverage has changed so drastically, I added some test cases but it looks like the hits and misses are changing abnormally? Did I miss something?

**Comment by Fokko** [comment]
@oliviersm199 It could be that you've worked from an old master.

**Comment by Fokko** [comment]
@jgao54 Any thoughts on this? :)

**Comment by oliviersm199** [comment]
Yes merging in master helped fix the code coverage report! Thank you @Fokko.

**Comment by oliviersm199** [comment]
Hello have the core committers had time to look at this? I know @Fokko requested input from @jgao54, let me know if you need anything from me.

**Comment by felipegasparini** [comment]
hey guys,

is there any ETA for this PR?

Btw, could you also update the documentation and release notes for 1.10 to clarify that plugins are not supported at this moment on the new RBAC UI? 
I put some effort on the migration to 1.10 only to rollback because of the lack of support for plugins.

File: tests/plugins/test_plugin.py [hunk 2]
```diff
@@ -67,17 +68,39 @@ def test(self):
...
+
 # Creating a flask blueprint to intergrate the templates and static folder
 bp = Blueprint(
     "test_plugin", __name__,
-    template_folder='templates',  # registers airflow/plugins/templates as a Jinja template folder
-    static_folder='static',
-    static_url_path='/static/test_plugin')
+    template_folder="templates",
+    static_folder="static",
+    static_url_path="/static/test_plugin")
```
Comment (ashb) [inline tests/plugins/test_plugin.py]: Style nit: please remove un-related changes like this from the PR.

File: tests/plugins/test_plugin.py [hunk 3]
```diff
@@ -58,7 +59,7 @@ def plugin_macro():
 
 # Creating a flask admin BaseView
 class TestView(BaseView):
-    @expose('/')
+    @expose("/")
```
Comment (ashb) [inline tests/plugins/test_plugin.py]: Style nit: please remove un-related changes like this from the PR.

File: docs/plugins.rst [hunk 4]
```diff
@@ -170,3 +191,16 @@ definitions in Airflow.
...
+        appbuilder_menu_items = [appbuilder_mitem]
+
+
+Note on role based views
+-------
+
+Airflow 1.10 introduced role based views using FlaskAppBuilder. You can configure which UI is used by setting rbac = True.
+To support plugin blueprints for both versions while maintaining backwards compatibility, additional fields in the
+AirflowTestPlugin class were added for specifically adding views and links to FlaskAppBuilder.
+
```
Comment (ashb) [inline docs/plugins.rst]: Remove empty lines.

File: docs/plugins.rst [hunk 5]
```diff
@@ -170,3 +191,16 @@ definitions in Airflow.
...
+        appbuilder_views = [v_appbuilder_package]
+        appbuilder_menu_items = [appbuilder_mitem]
+
+
+Note on role based views
+-------
+
+Airflow 1.10 introduced role based views using FlaskAppBuilder. You can configure which UI is used by setting rbac = True.
+To support plugin blueprints for both versions while maintaining backwards compatibility, additional fields in the
+AirflowTestPlugin class were added for specifically adding views and links to FlaskAppBuilder.
```
Comment (ashb) [inline docs/plugins.rst]: Please say here what the fields are.

File: docs/plugins.rst [hunk 6]
```diff
@@ -170,3 +191,16 @@ definitions in Airflow.
         admin_views = [v]
         flask_blueprints = [bp]
         menu_links = [ml]
+        appbuilder_views = [v_appbuilder_package]
+        appbuilder_menu_items = [appbuilder_mitem]
+
+
+Note on role based views
+-------
```
Comment (ashb) [inline docs/plugins.rst]: The `-----` should match the length of the line above otherwise RST will complain at us.

**Review by ashb — CHANGES_REQUESTED** [review]
Few minor nits then this looks good to merge.

**Comment by oliviersm199** [comment]
Made the requested changes.

File: docs/plugins.rst (hunk 6 above)
Comment (ashb) [reply docs/plugins.rst]: This one still needs doing I think?

Comment (oliviersm199) [reply docs/plugins.rst]: Hey I thought I changed it but must have made a mistake with git. Please check again :)

**Review by ashb — APPROVED** [review]

**Comment by ashb** [comment]
Tests failed due to flake8 style violation:

> ./tests/plugins/test_plugin.py:96:33: E261 at least two spaces before inline comment

File: tests/plugins/test_plugin.py [hunk 7]
```diff
@@ -67,17 +68,40 @@ def test(self):
...
+                    "category": "Search",
+                    "category_icon": "fa-th",
+                    "href": "https://www.google.com"}
+
+
 # Creating a flask blueprint to intergrate the templates and static folder
 bp = Blueprint(
     "test_plugin", __name__,
-    template_folder='templates',  # registers airflow/plugins/templates as a Jinja template folder
+    template_folder='templates', # registers airflow/plugins/templates as a Jinja template folder
```
Comment (ashb) [inline tests/plugins/test_plugin.py]: This double space is required by flake8.

**Comment by c4milo** [comment]
@ashb would it be possible to backport this PR into 1.10.x? It would be ideal for us.

**Comment by ashb** [comment]
@c4milo yeah it will be in 1.10.1 - check the "Fix Version" on a Jira to see what will be back-ported https://issues.apache.org/jira/browse/AIRFLOW-2744

**Comment by feng-tao** [comment]
@oliviersm199 , it seems that PluginRBACTest fails(currently it is disabled) if re-enable. Do you think you have time to fix the test?

**Comment by felipegasparini** [comment]
We also need to fix the example in the documentation (https://airflow.apache.org/plugins.html#example). It is currently broken: import fails, wrong method name in the view and missing path for templates.

I made a sample project to make this plugin integration work: https://github.com/felipegasparini/airflow_plugin_rbac_test/blob/dbaa049a9996df275b1d90f74b93ffbf206bb1d5/airflow/plugins/test_plugin/test_plugin.py

I will submit a PR to fix the doc later, but just posting it here since it may be useful for others.

**Pull request merged**